### PR TITLE
goroutine: only use named & error-logging background goroutines 

### PR DESCRIPTION
**Pull request by Strum355**
Previously, there were three ways to construct periodic background goroutines:
1. wrap a function with `goroutine.HandlerFunc`
2. create a type that implements `goroutine.Handler`
3. pass a function to `goroutine.NewHandlerWithErrorMessage`

Approach 1 had the behaviour where they would not log errors returned from the handler, which is of course not ideal : ) approach 2 has that too if not implementing `goroutine.ErrorHandler`

This PR creates a unified way of constructing them, where they all will log on error in a predetermined way if an customized way is not defined, and modernises it with `sourcegraph/log` instead of log15.

## Test plan

Nothing really to test, just swapping out Thing X thats proven to work for Thing Y thats proven to work but better :shrug:

### CODE REVIEW

File: enterprise/internal/insights/compression/worker.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func (i *CommitIndexer) Handler(ctx context.Context, observationCtx *observation
 	interval := time.Minute * time.Duration(intervalMinutes)
 
 	return goroutine.NewPeriodicGoroutineWithMetrics(ctx, interval,
-		goroutine.NewHandlerWithErrorMessage("commit_indexer_handler", func(ctx context.Context) error {
+		goroutine.NewHandlerWithErrorMessage("insights.commit_indexer_handler", "TODO", func(ctx context.Context) error {
```
Comment (Strum355) [inline enterprise/internal/insights/compression/worker.go:103]: @sourcegraph/code-insights-backend pls help me out here, what does this do? (and pls also proof check the other ones, I made a best attempt)

Comment (efritz) [inline enterprise/internal/insights/compression/worker.go:103]: TODO

Comment (Strum355) [reply enterprise/internal/insights/compression/worker.go:103]: pung up above lol

File: enterprise/internal/insights/background/insight_enqueuer.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ func newInsightEnqueuer(ctx context.Context, observationCtx *observation.Context
 	// See also https://github.com/sourcegraph/sourcegraph/pull/17227#issuecomment-779515187 for some very rough
 	// data retention / scale concerns.
 	return goroutine.NewPeriodicGoroutineWithMetrics(ctx, 1*time.Hour, goroutine.NewHandlerWithErrorMessage(
-		"insights_enqueuer",
+		"insights.enqueuer", "enqueues query runner and webhook runner jobs",
```
Comment (chwarwick) [inline enterprise/internal/insights/background/insight_enqueuer.go:43]:
```suggestion
		"insights.enqueuer", "enqueues snapshot and current recording query jobs",
```

**Review by efritz — APPROVED** [review]

File: enterprise/internal/insights/compression/worker.go (hunk 1 above)
Comment (chwarwick) [inline enterprise/internal/insights/compression/worker.go:103]:
```suggestion
		goroutine.NewHandlerWithErrorMessage("insights.commit_indexer_handler", "updates index of repo commits", func(ctx context.Context) error {
```

File: internal/goroutine/periodic.go [hunk 3]
```diff
@@ -57,41 +63,48 @@ func (f HandlerFunc) Handle(ctx context.Context) error {
...
+func NewPeriodicGoroutineWithMetrics(ctx context.Context, name, description string, interval time.Duration, handler Handler, operation *observation.Operation) *PeriodicGoroutine {
+	return newPeriodicGoroutine(ctx, name, description, interval, handler, operation, glock.NewRealClock())
 }
 
-func newPeriodicGoroutine(ctx context.Context, interval time.Duration, handler Handler, operation *observation.Operation, clock glock.Clock) *PeriodicGoroutine {
+func newPeriodicGoroutine(ctx context.Context, name, description string, interval time.Duration, handler Handler, operation *observation.Operation, clock glock.Clock) *PeriodicGoroutine {
 	ctx, cancel := context.WithCancel(ctx)
 
+	var h unifiedHandler
+	if uh, ok := handler.(unifiedHandler); ok {
```
Comment (efritz) [inline internal/goroutine/periodic.go:95]: uhhhh

Comment (Strum355) [reply internal/goroutine/periodic.go:95]: had me panicking there with that notification :gun:

**Review by efritz — APPROVED** [review]

**Pull request merged**